htpasswd/auth.go

```diff
@@ -16,64 +16,78 @@ import (
 	"time"
 )
 
-// Auth is the htpassword authenticater
-type Auth struct {
-	filename string
-	userHash map[string]string
+// File is a struct to serve an individual modTime
+type File struct {
+	name string
 	// Used in func reloadIfChanged to reload htpasswd file if it changed
 	modTime time.Time
-	mu      sync.RWMutex
 }
 
-// NewAuth creates an htpassword authenticater
-func NewAuth(filename string) (*Auth, error) {
-	a := &Auth{
-		filename: filename,
-	}
-	return a, a.parse(filename)
+// Auth is the htpassword authenticater
+type Auth struct {
+	filenames  []File
+	userHash   map[string]string
+	muUserHash sync.RWMutex
 }
 
-func (a *Auth) parse(filename string) error {
-	r, err := os.Open(filename)
-	if err != nil {
-		return err
+// NewAuth creates an htpassword authenticater
+func NewAuth(filenames []string) (*Auth, error) {
+	var htpasswdFiles []File
+	for _, file := range filenames {
+		htpasswdFiles = append(htpasswdFiles, File{name: file})
 	}
 
-	fileInfo, err := os.Stat(filename)
-	if err != nil {
-		return err
+	a := &Auth{
+		filenames: htpasswdFiles,
 	}
-	a.modTime = fileInfo.ModTime()
+	return a, a.parse(htpasswdFiles)
+}
 
-	cr := csv.NewReader(r)
-	cr.Comma = ':'
-	cr.Comment = '#'
-	cr.TrimLeadingSpace = true
+func (a *Auth) parse(filenames []File) error {
+	tmpUserHash := map[string]string{}
 
-	a.mu.Lock()
-	defer a.mu.Unlock()
-	a.userHash = map[string]string{}
-	for {
-		record, err := cr.Read()
-		if err == io.EOF {
-			break
+	for _, filename := range a.filenames {
+		r, err := os.Open(filename.name)
+		if err != nil {
+			return err
 		}
+
+		fileInfo, err := os.Stat(filename.name)
 		if err != nil {
 			return err
 		}
-		if len(record) != 2 {
-			return fmt.Errorf("password file in wrong format (%v)", filename)
+		filename.modTime = fileInfo.ModTime()
+
+		cr := csv.NewReader(r)
+		cr.Comma = ':'
+		cr.Comment = '#'
+		cr.TrimLeadingSpace = true
+
+		for {
+			record, err := cr.Read()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return err
+			}
+			if len(record) != 2 {
+				return fmt.Errorf("password file in wrong format (%v)", filename)
+			}
+			tmpUserHash[record[0]] = record[1]
 		}
-		a.userHash[record[0]] = record[1]
+		a.muUserHash.Lock()
+		a.userHash = tmpUserHash
+		a.muUserHash.Unlock()
```

Comment:
The replacing of the map has to be done outside of the for loop,
otherwise clients may read from partially filled maps.